examples/benchmarks/mnist.py

```diff
@@ -185,20 +190,29 @@ def train_torch(
     return model
 
 
-def train_step(model, dataset, device, criterion, batch_size, optimizer):
+def train_step(model, dataset, device, criterion, batch_size, optimizer, amp_enabled=False):
     model.train()
     y_preds = []
     losses = []
     batch_sizes = []
     tic = time.time()
+    scaler = None if not amp_enabled else torch.cuda.amp.GradScaler()
+
     for Xi, yi in torch.utils.data.DataLoader(dataset, batch_size=batch_size):
         Xi, yi = to_device(Xi, device), to_device(yi, device)
         optimizer.zero_grad()
-        y_pred = model(Xi)
-        y_pred = torch.log(y_pred)
-        loss = criterion(y_pred, yi)
-        loss.backward()
-        optimizer.step()
+        with torch.cuda.amp.autocast(enabled=amp_enabled):
+            y_pred = model(Xi)
+            y_pred = torch.log(y_pred)
+            loss = criterion(y_pred, yi)
+
+        if not amp_enabled:
+            loss.backward()
+            optimizer.step()
+        else:
+            scaler.scale(loss).backward()
+            scaler.step(optimizer)
+            scaler.update()
```

Comment:
Since PyTorch XLA requires to wrap the step of the optimizer as well (and this is a feature me might want to support in the future as well once TPUs become more accessible for smaller companies) I suggest that we introduce something akin to `self.optimizer_step(optimizer)` which sorts stuff like AMP scaling and XLA optimizations.

Reply:
I'd like to give this a try in the context of this PR to see if it makes things more ergonomic.